etc/contracts-test-data/evm-contracts/evm.sol

```diff
@@ -0,0 +1,328 @@
+        bytes memory data = bytes("test string.............................................................");
+        initialGas = gasleft();
+        hash = keccak256(data);
+        gasCost = initialGas - gasleft();
+        _requireGasCost(gasCost, 30 + 6 * (data.length + 31) / 32 + 10 /* additional PUSH* cost? */, "keccak256");
+        dummyHash = keccak256(abi.encode(dummyHash, value));
+    }
+
+    uint private constant GAS_TOLERANCE = 20;
+
+    function _requireGasCost(uint _actualCost, uint _expectedCost, string memory op) internal pure {
+        require(_actualCost >= _expectedCost, string.concat(op, " cost too low"));
+        require(_actualCost <= _expectedCost + GAS_TOLERANCE, string.concat(op, " cost too high"));
+    }
+
+    function testRecursion(bool _useFarCalls) external validEvmCall {
+        require(recurse(5, _useFarCalls) == 120, "recurse(5)");
+        require(recurse(10, _useFarCalls) == 3628800, "recurse(10)");
+    }
+
+    function recurse(uint _depth, bool _useFarCalls) public validEvmCall returns (uint) {
+        if (_useFarCalls) {
+            return (_depth <= 1) ? 1 : this.recurse(_depth - 1, _useFarCalls) * _depth;
+        } else {
+            return (_depth <= 1) ? 1 : recurse(_depth - 1, _useFarCalls) * _depth;
+        }
+    }
+
+    ICounter counter;
+
+    function testDeploymentAndCall(bytes32 _expectedCodeHash) external validEvmCall {
+        counter = new Counter(1);
+        require(address(counter) != address(0), "address(0)");
+        require(address(counter) != address(this), "address");
+        require(address(counter).codehash == _expectedCodeHash, "code hash");
+        require(address(counter).code.length != 0, "code");
+
+        require(counter.get() == 1, "counter.get()");
+        counter.increment(2);
+        require(counter.get() == 3, "counter.get() after");
+    }
+
+    function testCounterCall(uint _expectedInitialValue) external validEvmCall {
+        require(address(counter) != address(0), "counter not deployed");
+
+        require(counter.get() == _expectedInitialValue, "counter.get()");
+        counter.increment(3);
+        require(counter.get() == _expectedInitialValue + 3, "counter.get() after");
+
+        // Test catching reverts from EVM and EraVM contracts.
+        try counter.incrementWithRevert(4, true) {
+            revert("expected revert");
+        } catch Error(string memory reason) {
+            require(keccak256(bytes(reason)) == keccak256("This method always reverts"), "unexpected error");
+        }
+        require(counter.get() == _expectedInitialValue + 3, "counter.get() after revert");
+    }
+
+    uint private constant GAS_TO_SEND = 100000;
+
+    function testFarCalls(IGasTester _tester, bool _isEvmTester) external view validEvmCall {
+        uint gasMultiplier = _isEvmTester ? 1 : 5;
+        uint currentGas = gasleft();
+        if (_isEvmTester) {
+            // FIXME: doesn't work for EraVM tester (~115k less gas is passed)
```

Comment:
It should be decommit cost

Reply:
I've thought about it as well. Is this cost emulated by the EVM emulator? My confusion with this explanation is that it doesn't match how decommits are priced in EraVM; the amount of the subtracted gas doesn't depend on whether the called contact is "cold" or "hot" (i.e., was previously decommitted; IIUC, in the latter case, the decommit cost [should be 0](https://github.com/matter-labs/vm2/blob/457d8a7eea9093af9440662e33e598c13ba41633/crates/vm2/src/decommit.rs#L77-L83)). Note that *both* calls with default gas passed must be switched off; the second one calls a hot contract.

Reply:
Yes, this was implemented to somehow charge for decommits when we have EVM stipend (charge from callee frame). So it was supposed to charge even if we are calling already "hot" contract, because EVM doesn't know about hot/cold accounts in EraVM context.

But if we remove EVM stipend, we can remove this as well, I guess. Decommit cost will be charged from the caller frame